include/util/exception.hpp

```diff
@@ -54,6 +56,219 @@ class exception final : public std::exception
     virtual void anchor() const;
     const std::string message;
 };
+
+/**
+ * Indicates a class of error that occurred that was caused by some kind of
+ * external input (common examples are out of disk space, file permission errors,
+ * user supplied bad data, etc).
+ */
+class runtime_error : public std::runtime_error
+{
+    using Base = std::runtime_error;
+
+  protected:
+    explicit runtime_error(const char *message,
+                           const std::string &sourceref,
+                           const std::string &code_description)
+        : Base(message), sourceref(sourceref)
+    {
+        messagebuf = code_description + ": " + Base::what() + " (at " + sourceref + ")";
+    }
+    explicit runtime_error(std::string message,
+                           const std::string &sourceref,
+                           const std::string &code_description)
+        : runtime_error(message.c_str(), sourceref, code_description)
+    {
+    }
+
+  public:
+    const char *what() const noexcept override { return messagebuf.c_str(); }
+    const char *where() const noexcept { return sourceref.c_str(); }
+    virtual ErrorCode GetCode() const = 0;
+    virtual std::string GetCodeDescription() const = 0;
+
+  private:
+    // This function exists to 'anchor' the class, and stop the compiler from
+    // copying vtable and RTTI info into every object file that includes
+    // this header. (Caught by -Wweak-vtables under Clang.)
+    virtual void anchor() const;
+    const std::string sourceref; // Source location where the exception was raised
+    std::string messagebuf;
+};
+
+// TODO: the classes below  are dying for templatization, but I don't know of a good pattern to
+// instantiate a template with string literals that isn't hella ugly.
```

Comment:
There isn't if you don't want to do workarounds such as

https://gist.github.com/daniel-j-h/6bc186ff832647bb128895534fce61bb